jumanji/environments/logic/sliding_tile_puzzle/env.py

```diff
@@ -0,0 +1,266 @@
+from jax import lax
+from numpy.typing import NDArray
+
+from jumanji import specs
+from jumanji.env import Environment
+from jumanji.environments.logic.sliding_tile_puzzle.generator import (
+    Generator,
+    RandomGenerator,
+)
+from jumanji.environments.logic.sliding_tile_puzzle.reward import (
+    ImprovedDenseRewardFn,
+    RewardFn,
+)
+from jumanji.environments.logic.sliding_tile_puzzle.types import Observation, State
+from jumanji.environments.logic.sliding_tile_puzzle.viewer import (
+    SlidingTilePuzzleViewer,
+)
+from jumanji.types import TimeStep, restart, termination, transition
+from jumanji.viewer import Viewer
+
+
+class SlidingTilePuzzle(Environment[State]):
+    """Environment for the Sliding Tile Puzzle problem.
+
+    The problem is a combinatorial optimization task where the goal is
+    to move the empty tile around in order to arrange all the tiles in order.
+
+    - observation: `Observation`
+        - puzzle: jax array (int32) of shape (N, N), representing the current state of the puzzle.
+        - empty_tile_position: Tuple of int32, representing the position of the empty tile.
+        - action_mask: jax array (bool) of shape (4,), indicating which actions are valid
+            in the current state of the environment.
+
+    - action: Tuple of int32, representing the direction to move the empty tile
+        (up, down, left, right)
+
+    - reward: float, a dense reward is provided based on the arrangement of the tiles.
+        It equals the negative sum of the boolean difference between
+        the current state of the puzzle and the goal state (correctly arranged tiles).
+        Each incorrectly placed tile contributes -1 to the reward.
+
+    - episode termination: if the puzzle is solved.
+
+    - state: `State`
+        - puzzle: jax array (int32) of shape (N, N), representing the current state of the puzzle.
+        - empty_tile_position: Tuple of int32, representing the position of the empty tile.
+        - key: jax array (uint32) of shape (2,), random key used to generate random numbers
+            at each step and for auto-reset.
+    """
+
+    def __init__(
+        self,
+        generator: Optional[Generator] = None,
+        reward_fn: Optional[RewardFn] = None,
+        viewer: Optional[Viewer[State]] = None,
+    ) -> None:
+        """Instantiate a `SlidingTilePuzzle` environment.
+
+        Args:
+            generator: callable to instantiate environment instances.
+                Defaults to `RandomGenerator` which generates puzzles with
+                a size of 5x5.
+            reward_fn: RewardFn whose `__call__` method computes the reward of an environment
+                transition. The function must compute the reward based on the current state,
+                the chosen action and the next state.
+                Implemented options are [`DenseRewardFn`, `SparseRewardFn`,
+                `ImprovedDenseRewardFn`]. Defaults to `ImprovedDenseRewardFn`.
```

Comment:
I think we should only have dense and sparse, ImprovedDense is a bit confusing.

```suggestion
                Implemented options are [`DenseRewardFn`, `SparseRewardFn`]. Defaults to `DenseRewardFn`.
```